### ISSUE TITLE
Optimisations to context creation

### ISSUE THREAD

**Issue by tristanic** [openmm/openmm#2474]
With all the recent improvements to simulation preparation in OpenMM 7.4.0, the time for starting a new simulation in ISOLDE is down to 1-3 seconds in typical use cases (~200 residues, implicit solvent, MDFF, CMAP, and a half-dozen custom forces). Which is fantastic, and greatly improves usability for an interactive environment that involves constantly stopping and starting lots of small local simulations.

I'm still curious about what more can be done here, though.  Profiling startup of a fairly typical simulation gives me the following:

```
         334589 function calls (330304 primitive calls) in 2.173 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    1.569    1.569    1.569    1.569 {built-in method _openmm.new_Context}
        1    0.039    0.039    0.087    0.087 forcefield.py:2030(createForce)
        1    0.030    0.030    0.237    0.237 forcefield.py:1026(createSystem)
        1    0.029    0.029    0.033    0.033 forcefield.py:1925(createForce)
```
... showing that preparing the kernels and finalising the `Context` is now the rate-limiting step. It's also surprisingly variable - starting the same simulation multiple times over shows the time required for `new_Context()` varying by about +/- 0.5 seconds. I'm wondering if there are any possibilities for easy gains here? Some thoughts:

- could kernel preparation be parallelised over the available CPU cores?
- Is there scope to cache the Lepton `CompiledExpression`(s) for a given custom force so they only need to be re-computed if the energy function changes? For example, could the Python session maintain a dict of {energy function hash : Lepton object} so each function only needs to be parsed at first use? 

A case in point for the second one: for MDFF I have two different `CustomCompoundBondForce` implementations. The first just uses `Continuous3DFunction` for the map interpolation. The second is designed for large maps where the `Continuous3DFunction` array size gets too big for the GPU, and instead performs the cubic interpolation on-the-fly - so all that's stored on the GPU is a `Discrete3DFunction` containing the map grid points. Obviously makes for a much more complex energy function (although surprisingly the performance difference is barely noticeable once the simulation is running). But the difference in complexity shows in simulation startup time: if I force the use of this lower-memory-usage implementation for the same test case as timed above, then the `new_Context()` call time increases to about 4.5 seconds.

No real pressure - all these timings are already great for almost all tasks.

**Comment by peastman** [comment]
Assuming no other bottlenecks, the minimum time for Context creation varies widely between platforms.  It typically is in the range of tens of milliseconds for Reference, hundreds of milliseconds for CPU, and seconds for CUDA and OpenCL.  There are other sources of variation, though.  NVIDIA's OpenCL starts up much faster than AMD's, I think because it caches compiled kernels from previous runs.  Compiling CUDA kernels is *really* slow, about one second for every compilation unit.  We do our own caching of compiled CUDA kernels, so the first time it will be very slow, but then if you create another identical Context it will be much faster.

But there can be other things that make particular Contexts very slow to create.  Compiling Lepton expressions is very fast, so I doubt that's an issue.  Building splines for a large Continuous3DFunction can take a while.  Preparing large numbers of connected constraints can also be slow, since it involves inverting a matrix.

So the first thing to do is try different platforms.  If you use Reference does context creation become fast?  (Actually simulating anything will be very slow of course, but I'm just asking about the time to create the context.)  If so, you're probably just seeing the overhead of CUDA or OpenCL, and there isn't a lot we can do about it.  But if creating a reference context is still slow, then it's something else.

**Comment by tristanic** [comment]
Hmm... on the Reference platform context creation is actually much slower:

```
         1134168 function calls (1124772 primitive calls) in 36.667 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   35.041   35.041   35.041   35.041 {built-in method _openmm.new_Context}
```
... and essentially the same (as expected, I suppose) for the CPU platform. If I switch to the low-memory MDFF implementation that blows out further to 66s. 

Can't test CUDA on my home machine (Fedora 29 - GCC version is too new), but my past experience is that it is as you say a bit slower than OpenCL to prepare the context (but significantly faster in actual simulation performance).

**Comment by tristanic** [comment]
Oh, and I don't have any connected constraints - hydrogens are rigid, but that's it.

**Comment by peastman** [comment]
That's odd.  I really don't know what the bottleneck is then.  I don't see why processing the tabulated function would be different.  On both platforms, it mostly comes down to a call to `SplineFitter::create3DNaturalSpline()`.  It would definitely be worth profiling and/or instrumenting the code to identify where the time is being spent.

**Comment by tristanic** [comment]
Will see what I can find out (when time permits). At this stage it would be really difficult to extract a self-contained test case out of ChimeraX. Can send you the MDFF force implementations if you're interested (they may be of general interest to other users, and are quite thoroughly play-tested at this stage).

**Comment by peastman** [comment]
Is it implemented with a plugin or with custom forces?  If the latter, you could just serialize the System to an XML file and send that.

**Comment by tristanic** [comment]
With custom forces. I haven't used OpenMM's serialisation before - will have a look tomorrow.

**Comment by peastman** [comment]
```python
with open('system.xml', 'w') as f:
  f.write(XmlSerializer.serialize(system))
```

That's all there is to it.

**Comment by tristanic** [comment]
You're right - that _is_ easy!

OK, I've put XML files using the two different MDFF implementations at the following links:

https://drive.google.com/open?id=1sKQsta6sDPcM81V-2-UlVo3hYQKOX9Ql
https://drive.google.com/open?id=1vXfRbwGlS2NXZy-6pzd1rTHXMAP_b2xV

**Comment by peastman** [comment]
Great, I'll take a look.

Reading files back in is just as easy:

```python
with open('system.xml') as f:
    system = XmlSerializer.deserialize(f.read())
```

**Comment by tristanic** [comment]
Just profiled the same system on my CentOS 7 box with CUDA 10.1. Seems things have improved a lot since last time I checked: startup time is now faster (and much more consistent) than OpenCL (about 1 +/- 0.1 seconds for `new_Context()`, and actual simulation performance is markedly faster. 

Up until now I've had OpenCL set as the default platform because of (a) startup times and (b) CUDA not supporting the default GCC versions in bleeding-edge distros - but it seems both of those issues are now in the past. Seems pretty clear now that CUDA is the superior choice on all criteria for systems that have it.

**Comment by tristanic** [comment]
... but looking at it a bit more, that's only the case on repeated starts when the topology is identical every time (i.e. when the whole structure is modelled). That's rarely the case for ISOLDE, where the subset of residues that go into the simulation topology is defined based on whatever the user has selected at the time. If I pick a single random atom (which based on ISOLDE's rules expands to a simulated topology of ~2000 atoms), then `new_Context() takes about 3 seconds for CUDA and about 1.5 for OpenCL. So the new question is, is there any scope for acceleration when the contingent of forces in each simulation is identical, but the number and identity of particles within each force changes?

**Comment by peastman** [comment]
It turns out I was wrong: in the case of the reference platform, compiling the Lepton expressions really is the bottleneck.  I don't know why it's as slow as it is, but I plan to investigate.  The slowness is entirely from the CustomCompoundBondForce.  The expression isn't that complicated, and it only uses one particle per bond, so that seems very strange.

The CUDA behavior is from compiling kernels.  If it has cached versions of everything it's fast, but if the number of particles has changed it has to recompile some of them.  I don't see anything I can do about that short of passing in the number of particles as an argument rather than compiling it into the kernel.  But that would trade faster startup time for slower execution.    I already try to do that for kernels that aren't performance sensitive.

**Comment by tristanic** [comment]
I have a little Twitter poll of my current users going on, and so far 6 out 7 voters are happy to trade off the slightly slower startup time for better simulation performance. :)

> I don't see anything I can do about that short of passing in the number of particles as an argument rather than compiling it into the kernel. But that would trade faster startup time for slower execution.

If it doesn't horribly complicate everything, I think there would be value in making this optional. Of course for big, long-running simulations raw performance is by far the most important. But when model building into density there are many scenarios involving tiny and very short simulations - e.g. changing a rotamer or flipping a peptide bond then settling the local environment, or adding then minimising a new residue. For those, 4 seconds of initialisation time can easily be longer than the actual simulation time needed, so bringing it down to sub-second levels would be of real value. Having the ability to use precompiled variable-length kernels for those then switch back to the default just-in-time compilation when working on large fragments would be fantastic.

**Comment by peastman** [comment]
Something really weird is going on with the reference code.  The reference and OpenCL platforms have basically identical code.  Each takes an identical ParsedExpression and calls `differentiate()` to compute the derivative with respect to the same variable.  In the OpenCL platform, it takes a tiny fraction of a second.  For the Reference platform, it takes multiple seconds to complete.

**Comment by tristanic** [comment]
Incidentally, this explains why falling back to CPU during minimisation was proving so lethal to my simulation performance.

**Comment by peastman** [comment]
I figured it out.  The difference is coming from the tabulated function.  While differentiating and optimizing the expression, it keeps having to clone nodes in the tree.  Including the one representing the function.  On the Reference platform that's a ReferenceContinuous3DFunction, a fairly heavyweight object containing a complete copy of the function and the spline coefficients.  On the OpenCL platform it's a very lightweight placeholder object.

That should be easy to improve.  Since these objects are immutable, I can create a lightweight wrapper class that's just a shared pointer to the real function implementation.  Cloning it then just means copying the pointer.

**Comment by peastman** [comment]
The fix is in #2476.  The Context creation time with the Reference platform is now less than a second.
